lib/ios-deploy.js

```diff
@@ -72,14 +72,21 @@ class IOSDeploy {
         } else {
           const readStream = fs.createReadStream(itemPath, {autoClose: true});
           const writeStream = await afcService.createWriteStream(pathOnPhone, {autoDestroy: true });
-          promises.push(new B((resolve) => writeStream.on('close', resolve)));
+          promises.push(new B((resolve, reject) => {
+            writeStream.on('error', reject);
+            writeStream.on('close', resolve);
+          }).reflect());
           readStream.pipe(writeStream);
         }
       });
       try {
-        await B.all(promises).timeout(APPLICATION_PUSH_TIMEOUT);
+        await B.all(promises).timeout(APPLICATION_PUSH_TIMEOUT).each(p => {
```

Comment:
does `each` exist here?

Reply:
Yeap. It gives you another promise which is finished when all promises are iterated

Reply:
Is this `each` loop really necessary? I don't think that it will ever get run if there is an error in any of the promises:
```js
try {
  await B.all([B.resolve('a'), B.reject('b')]).timeout(500).each((p) => {
    console.log('p:', p);
    if (p.isRejected()) {
      console.log('error, throwing');
      throw p.reason();
    }
  });
} catch (err) {
  console.log('err:', err);
}

// output:
err: b
```

In the event of a timeout error it isn't run either:
```js
try {
  await B.all([B.resolve('a'), B.resolve('b'), B.delay(1000)]).timeout(500).each((p) => {
    console.log('p:', p);
    if (p.isRejected()) {
      console.log('error, throwing');
      throw p.reason();
    }
  });
} catch (err) {
  console.log('err:', err);
}

// output:
err: TimeoutError: operation timed out
    at afterTimeout (/Users/isaac/code/appium-xcuitest-driver/node_modules/bluebird/js/release/timers.js:46:19)
    at Timeout.timeoutTimeout [as _onTimeout] (/Users/isaac/code/appium-xcuitest-driver/node_modules/bluebird/js/release/timers.js:76:13)
    at ontimeout (timers.js:436:11)
    at tryOnTimeout (timers.js:300:5)
    at listOnTimeout (timers.js:263:5)
    at Timer.processTimers (timers.js:223:10)
```

Reply:
@imurchie You are right. I'm thinking how to rewrite this. I'm trying to make sure every promise is finished so the resources are closed properly

Reply:
@umutuzgur Actually, I missed the fact that you call `reflect` on each promise. So they will not themselves throw, and my problem is gone.

The only problem remains the timeout, which currently would throw outside of the individual promises. But I think you could get around that by setting the timeout on the individual promises, before reflecting.
```js
promises.push(new B((resolve, reject) => {
  writeStream.on('error', reject);
  writeStream.on('close', resolve);
}).timeout(APPLICATION_PUSH_TIMEOUT).reflect());
```
Then do the `each` loop as above. I think it should collect the errors though (in case there a more) rather than throwing the first.